File: 2025/src/bin/day4/day4_p2.rs
use std::fs;

fn main() {

    fn add(x: usize, y: i32) -> usize {
        if y.is_negative() {
            x - y.wrapping_abs() as u32 as usize
        } else {
            x + y as usize
        }
    }

    fn print(grid: Vec<Vec<char>>) {
        for line in grid.iter() {
            let s: String = line.into_iter().collect();
            println!("{:?}", s);
        }
    }

    let contents = fs::read_to_string("src/bin/day4/data.txt")
        .expect("Should have been able to read the file");

    let mut grid: Vec<_> = contents.lines().map(|line| line.chars().collect::<Vec<_>>()).collect();
    let mut prev_total = -1;
    let mut total = 0;

    fn remove(mut grid: Vec<Vec<char>>, mut total: i32) -> (Vec<Vec<char>>, i32) {
        let rows: usize = grid.len();
        let cols: usize = grid[0].len();
        
        for r in 0..rows {
            for c in 0..cols {
                if grid[r][c] == '@' {
                    let dirs: Vec<(i32, i32)> = vec![(-1, 0), (-1, 1), (-1, -1), (0, 1), (0, -1), (1, 0), (1, 1), (1, -1)];
                    let mut neighbours = 0;
                    for (dx, dy) in dirs {
                        if dx.is_negative() {
                            if r < dx.wrapping_abs() as u32 as usize {
                                continue
                            }
                        }
                        if dy.is_negative() {
                            if c < dy.wrapping_abs() as u32 as usize {
                                continue
                            }
                        }
                        let new_r = add(r, dx);
                        let new_c = add(c, dy);
                        if new_r < rows && new_c < cols {
                            if grid[new_r][new_c] == '@' {
                                neighbours += 1;
                            }
                        }

                    }
                    if neighbours < 4 {
                        total += 1;
                        grid[r][c] = 'x';
                    } 
                }
            }
        };
        (grid, total)
    }
    while total > prev_total {
        prev_total = total;
        (grid, total) = remove(grid, total);
    }
    
    print(grid);
    println!("{:?}", total);
}

File: 2022/day4/day4_p2.py
def main():
    with open("day4/data.txt", "r") as f:
        contents = f.read()
    

        lines = contents.splitlines()
        result = 0
        for line in lines:
            x, y = line.split(',')
            x_1, x_2 = x.split('-')
            y_1, y_2 = y.split('-')
            
            x_1, x_2 = int(x_1), int(x_2)
            y_1, y_2 = int(y_1), int(y_2)

            if x_1 == y_1 or x_2 == y_2:
                result += 1
            elif x_1 <= y_1 and x_2 >= y_2:
                result +=1 
            elif y_1 <= x_1 and y_2 >= x_2:
                result += 1
            elif y_1 <= x_1 and x_1 <= y_2:
                result += 1 
            elif y_1 <= x_2 and x_2 <= y_2:
                result += 1 
            elif x_1 <= y_1 and y_1 <= x_2:
                result += 1 
            elif x_1 <= y_2 and y_2 <= x_2:
                result += 1 

        print(result)
        
if __name__ == "__main__":
    main()
